diff --git a/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetValuesReader.java b/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetValuesReader.java
index 116e23fb..6014bac5 100644
--- a/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetValuesReader.java
+++ b/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedParquetValuesReader.java
@@ -103,18 +103,26 @@ public final class VectorizedParquetValuesReader extends BaseVectorizedParquetVa
           bufferIdx += numValues;
           break;
         case PACKED:
-          for (int i = 0; i < numValues; ++i) {
-            if (packedValuesBuffer[packedValuesBufferIdx++] == maxDefLevel) {
-              vector.getDataBuffer().setLong(bufferIdx * typeWidth, valuesReader.readLong());
-              if (setArrowValidityVector) {
+          if (setArrowValidityVector) {
+            for (int i = 0; i < numValues; ++i) {
+              if (packedValuesBuffer[packedValuesBufferIdx++] == maxDefLevel) {
+                vector.getDataBuffer().setLong(bufferIdx * typeWidth, valuesReader.readLong());
                 BitVectorHelper.setValidityBitToOne(vector.getValidityBuffer(), bufferIdx);
               } else {
+                setNull(nullabilityHolder, bufferIdx, vector.getValidityBuffer());
+              }
+              bufferIdx++;
+            }
+          } else {
+            for (int i = 0; i < numValues; ++i) {
+              if (packedValuesBuffer[packedValuesBufferIdx++] == maxDefLevel) {
+                vector.getDataBuffer().setLong(bufferIdx * typeWidth, valuesReader.readLong());
                 nullabilityHolder.setNotNull(bufferIdx);
+              } else {
+                setNull(nullabilityHolder, bufferIdx, vector.getValidityBuffer());
               }
-            } else {
-              setNull(nullabilityHolder, bufferIdx, vector.getValidityBuffer());
+              bufferIdx++;
             }
-            bufferIdx++;
           }
           break;
       }
