diff --git a/versions.props b/versions.props
index 6f8723cc..5a0ab857 100644
--- a/versions.props
+++ b/versions.props
@@ -1,7 +1,7 @@
 org.slf4j:slf4j-api = 1.7.22
 com.google.guava:guava = 28.0-jre
 org.apache.avro:avro = 1.8.2
-org.apache.arrow:arrow-vector = 0.14.1
+org.apache.arrow:arrow-vector = 0.15.0
 org.apache.hadoop:* = 2.7.3
 org.apache.hive:hive-standalone-metastore = 1.2.1
 org.apache.orc:orc-core = 1.5.6
